### PR TITLE
Use maxReadLevel to read task in source cluster when initialize replication

### PR DESCRIPTION
**Pull request by hai719**
## What changed?
Use current queueExclusiveHighReadWatermark to send catch up tasks when connections between clusters are new.

## Why?
When a new pair of clusters is connected, we don't want sender to sync all previous tasks. Only new replication tasks are needed to be synced to receiver.

## How did you test it?
Tested locally by checking the newly added logs.

### CODE REVIEW

**Review by xwduan — APPROVED** [review]

File: service/history/replication/stream_receiver.go [hunk 1]
```diff
@@ -160,14 +161,21 @@ func (r *StreamReceiverImpl) sendEventLoop() error {
...
 		select {
 		case <-timer.C:
 			timer.Reset(r.Config.ReplicationStreamSyncStatusDuration())
-			if err := r.ackMessage(r.stream); err != nil {
+			watermark, err := r.ackMessage(r.stream)
+			if err != nil {
 				r.logger.Error("StreamReceiver exit send loop", tag.Error(err))
 				return err
 			}
+			if watermark != inclusiveLowWatermark {
```
[the remark this answers is not in the document]
Comment (tdeebswihart) [reply service/history/replication/stream_receiver.go:175]: How often does it change? Even if we're not emitting the logs it takes time to format and process them

File: service/history/replication/stream_receiver.go [hunk 2]
```diff
@@ -246,10 +254,12 @@ func (r *StreamReceiverImpl) processMessages(
 		)
 		exclusiveHighWatermark := streamResp.Resp.GetMessages().ExclusiveHighWatermark
 		exclusiveHighWatermarkTime := timestamp.TimeValue(streamResp.Resp.GetMessages().ExclusiveHighWatermarkTime)
+		r.logger.Debug(fmt.Sprintf("StreamReceiver processMessages received %d tasks with exclusiveHighWatermark %d", len(tasks), exclusiveHighWatermark))
 		for _, task := range r.taskTracker.TrackTasks(WatermarkInfo{
 			Watermark: exclusiveHighWatermark,
 			Timestamp: exclusiveHighWatermarkTime,
 		}, tasks...) {
+			r.logger.Debug("StreamReceiver processMessages submit task to scheduler", tag.TaskID(task.TaskID()))
```
Comment (xwduan) [inline service/history/replication/stream_receiver.go:262]: This is probably unnecessary.

File: service/history/replication/stream_receiver.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (xwduan) [reply service/history/replication/stream_receiver.go:175]: Good point. This should change very often. Then I think the best practice is to try not emit debug log to save cpu cycle?

File: service/history/replication/stream_receiver.go [hunk 3]
```diff
@@ -191,11 +199,11 @@ func (r *StreamReceiverImpl) recvEventLoop() error {
 
 func (r *StreamReceiverImpl) ackMessage(
```
Comment (tdeebswihart) [inline service/history/replication/stream_receiver.go:200]: Can you make a note on this function that it returns the watermark info if present?

File: service/history/replication/stream_sender.go [hunk 4]
```diff
@@ -354,6 +365,7 @@ Loop:
 		if task == nil {
 			continue Loop
 		}
+		s.logger.Debug("StreamSender send replication task", tag.TaskID(task.SourceTaskId))
```
Comment (tdeebswihart) [inline service/history/replication/stream_sender.go:368]: Do we need to log every task?

File: service/history/replication/stream_sender.go [hunk 5]
```diff
@@ -256,21 +263,24 @@ func (s *StreamSenderImpl) sendCatchUp() (int64, error) {
 		s.clientShardKey.ShardID,
 	)
 
+	catchupEndExclusiveWatermark := s.shardContext.GetQueueExclusiveHighReadWatermark(tasks.CategoryReplication).TaskID
```
Comment (yycptt) [inline service/history/replication/stream_sender.go:266]: when will QueueExclusiveHighReadWatermark be updated/persisted for replication queue?

**Review by yycptt — COMMENTED** [review]
Not quite sure I follow. QueueExclusiveHighReadWatermark meant to be the maxReadLevel, i.e. beyond which task can NOT be read by the processor, which is different from the ack level concept.

**Review by yycptt — APPROVED** [review]

**Comment by hai719** [comment]
> ~Not quite sure I follow. QueueExclusiveHighReadWatermark meant to be the maxReadLevel, i.e. beyond which task can NOT be read by the processor, which is different from the ack level concept.~
> 
> Please ignore this comment. I confused GetQueueExclusiveHighReadWatermark with the one in queueState when reviewing. (I think the PR title should be use maxReadLevel to read task..., instead of ack level.)

Title updated.

**Review by yux0 — APPROVED** [review]

**Review by hai719 — COMMENTED** [review]
Just found out that I forgot to submit my comments addressing your comments for this PR during review. Sorry for the confusions.

**Pull request merged**